File: crates/bridges/dukascopy/src/proto.rs
//! The JSON-lines stdio protocol between `DukascopyExecutionClient` and the JForex sidecar.
//!
//! One compact JSON object per line. Rust → sidecar: [`Command`] (tag `cmd`).
//! Sidecar → Rust: [`Envelope`] (tag `kind`); the `event` payload is the canonical
//! [`Event`] tagged union, so the Rust side needs no venue mapping at all. Parse
//! failures return `None` — callers log-and-skip (forward compatibility, never a
//! crash on a bad line). Both directions derive Serialize AND Deserialize so the
//! fake test bridge can speak the protocol in reverse.

use serde::{Deserialize, Serialize};
use vike_model::events::Event;
use vike_model::OrderRequest;

/// Rust → sidecar command. `order` is the `OrderRequest` serde shape verbatim —
/// `crates/vike-model/src/order.rs` is the single schema authority for both sides.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "cmd", rename_all = "lowercase")]
pub enum Command {
    Submit { order: Box<OrderRequest> },
    Cancel { client_order_id: String },
    Shutdown,
}

/// Sidecar → Rust envelope. Exactly one of `ready` | `fatal` opens the stream
/// (the handshake); `event` lines carry the canonical [`Event`].
///
/// `position` (additive, netting-truth law A7): the sidecar's AUTHORITATIVE per-instrument net
/// position — signed size in units and the signed-amount-weighted average of its remaining
/// per-order open prices — emitted right after every fill it reports. JForex is
/// position-per-order and the sidecar realizes netted closes at EACH order's own entry, so the
/// Rust blended fold can drift in (realized, avg) while net size agrees; this line carries the
/// venue truth the exec client re-anchors against (see `netting.rs`). Back-compatible both
/// directions: an old Rust reader sees an unknown `kind` and logs-and-skips; a new reader with
/// an old jar simply never receives the line (no re-anchor — pre-fix behavior).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "lowercase")]
pub enum Envelope {
    Ready {
        account: String,
        #[serde(default)]
        balance: f64,
    },
    Event {
        event: Box<Event>,
    },
    Position {
        /// Canonical symbol (EURUSD form), same as `FillEvent.symbol`.
        symbol: String,
        /// Signed net size in UNITS (not JForex millions): >0 long, <0 short, 0 flat.
        size: f64,
        /// Signed-amount-weighted average open price of the remaining orders; 0.0 when flat.
        avg_px: f64,
        #[serde(default)]
        ts: i64,
    },
    Fatal {
        reason: String,
    },
}

/// Compact one-line JSON (no trailing newline — writers add it).
pub fn encode_line<T: Serialize>(msg: &T) -> String {
    serde_json::to_string(msg).expect("protocol types always serialize")
}

/// Parse a sidecar stdout line; `None` = unparseable/unknown (caller logs + skips).
pub fn parse_envelope(line: &str) -> Option<Envelope> {
    serde_json::from_str(line).ok()
}

/// Parse a Rust command line (used by the fake bridge; the Java twin is `Proto.java`).
pub fn parse_command(line: &str) -> Option<Command> {
    serde_json::from_str(line).ok()
}

#[cfg(test)]
mod tests {
    use super::*;
    use vike_model::events::OrderAccepted;

    fn order(coid: &str) -> OrderRequest {
        OrderRequest {
            combo_legs: Vec::new(),
            client_order_id: coid.into(),
            venue: "dukascopy".into(),
            symbol: "EURUSD".into(),
            side: 1,
            qty: 1000.0,
            order_type: "market".into(),
            price: None,
            trigger_price: None,
            reduce_only: false,
            time_in_force: Default::default(),
            gtd_expiry: None,
            ts: 42,
            parent_order_id: None,
            linked_order_ids: vec![],
            order_list_id: None,
            contingency_type: None,
            weight: 0.0,
            stop: None,
            trail: None,
            extreme: None,
            on_close: false,
            margin_mode: None,
            trigger_by: None,
        }
    }

    #[test]
    fn command_round_trips_and_uses_wire_names() {
        let cmd = Command::Submit { order: Box::new(order("c1")) };
        let line = encode_line(&cmd);
        // wire contract: tag + real OrderRequest field names (order_type, not "kind")
        assert!(line.contains(r#""cmd":"submit""#), "{line}");
        assert!(line.contains(r#""order_type":"market""#), "{line}");
        assert!(line.contains(r#""client_order_id":"c1""#), "{line}");
        assert!(!line.contains('\n'));
        assert_eq!(parse_command(&line), Some(cmd));

        let cancel = Command::Cancel { client_order_id: "c1".into() };
        let line = encode_line(&cancel);
        assert!(line.contains(r#""cmd":"cancel""#), "{line}");
        assert_eq!(parse_command(&line), Some(cancel));

        assert_eq!(parse_command(&encode_line(&Command::Shutdown)), Some(Command::Shutdown));
    }

    #[test]
    fn envelope_round_trips_with_canonical_event() {
        let env = Envelope::Event {
            event: Box::new(Event::OrderAccepted(OrderAccepted {
                client_order_id: "c1".into(),
                venue_order_id: Some("12345".into()),
                ts: 9,
            })),
        };
        let line = encode_line(&env);
        assert!(line.contains(r#""kind":"event""#), "{line}");
        assert!(line.contains(r#""type":"OrderAccepted""#), "{line}");
        assert_eq!(parse_envelope(&line), Some(env));

        let ready = parse_envelope(r#"{"kind":"ready","account":"DEMO2cGyrc","balance":100000.0}"#);
        assert_eq!(
            ready,
            Some(Envelope::Ready { account: "DEMO2cGyrc".into(), balance: 100000.0 })
        );
        // balance is optional (serde default)
        assert_eq!(
            parse_envelope(r#"{"kind":"ready","account":"A"}"#),
            Some(Envelope::Ready { account: "A".into(), balance: 0.0 })
        );
        let fatal = parse_envelope(r#"{"kind":"fatal","reason":"login failed"}"#);
        assert_eq!(fatal, Some(Envelope::Fatal { reason: "login failed".into() }));
    }

    /// The additive `position` envelope: round-trips, uses the wire names the Java emitter
    /// writes, and its ABSENCE keeps every pre-existing envelope parsing identically — the
    /// serde back-compat pin for the netting-truth protocol extension (both directions: the
    /// old-reader side is pinned by `bad_lines_parse_to_none`'s unknown-kind → `None` law).
    #[test]
    fn position_envelope_round_trips_and_is_additive() {
        let env =
            Envelope::Position { symbol: "EURUSD".into(), size: -1_000_000.0, avg_px: 1.1, ts: 42 };
        let line = encode_line(&env);
        assert!(line.contains(r#""kind":"position""#), "{line}");
        assert!(line.contains(r#""symbol":"EURUSD""#), "{line}");
        assert_eq!(parse_envelope(&line), Some(env));

        // Exactly what the Java emitter writes (Gson property order), ts optional.
        let java = r#"{"kind":"position","symbol":"USDJPY","size":2000.0,"avg_px":155.25,"ts":7}"#;
        assert_eq!(
            parse_envelope(java),
            Some(Envelope::Position {
                symbol: "USDJPY".into(),
                size: 2000.0,
                avg_px: 155.25,
                ts: 7
            })
        );
        assert_eq!(
            parse_envelope(r#"{"kind":"position","symbol":"USDJPY","size":0.0,"avg_px":0.0}"#),
            Some(Envelope::Position { symbol: "USDJPY".into(), size: 0.0, avg_px: 0.0, ts: 0 })
        );
    }

    #[test]
    fn bad_lines_parse_to_none() {
        assert_eq!(parse_envelope("not json"), None);
        assert_eq!(parse_envelope(r#"{"kind":"mystery"}"#), None);
        assert_eq!(parse_command(r#"{"cmd":"mystery"}"#), None);
        assert_eq!(parse_command(""), None);
    }
}

File: crates/bridges/dukascopy/jforex-bridge/src/main/java/vike/jforex/Proto.java
package vike.jforex;

import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import java.io.PrintStream;

/**
 * The stdio protocol: parse Rust commands from stdin lines, write envelope lines to
 * stdout. ALL stdout writes go through one synchronized writer (spec: interleaved
 * thread writes must never corrupt a line — a corrupted line is a silently dropped
 * fill). stderr is for humans; never write protocol there.
 */
final class Proto {
    private final PrintStream out;

    Proto(PrintStream out) {
        this.out = out;
    }

    /** Parse a stdin line; null = unparseable/unknown (caller logs + skips, never exits).
     *  `cmd` must be a string primitive — `{"cmd":{...}}` or `{"cmd":123}` is null too. */
    static JsonObject parseCommand(String line) {
        try {
            var el = JsonParser.parseString(line);
            if (!el.isJsonObject()) return null;
            JsonObject obj = el.getAsJsonObject();
            var cmd = obj.get("cmd");
            if (cmd == null || !cmd.isJsonPrimitive() || !cmd.getAsJsonPrimitive().isString()) {
                return null;
            }
            return obj;
        } catch (RuntimeException e) {
            return null;
        }
    }

    synchronized void writeLine(JsonObject envelope) {
        out.println(envelope); // Gson JsonObject#toString is compact single-line JSON
        out.flush();
    }

    void ready(String account, double balance) {
        JsonObject o = new JsonObject();
        o.addProperty("kind", "ready");
        o.addProperty("account", account);
        o.addProperty("balance", balance);
        writeLine(o);
    }

    void fatal(String reason) {
        JsonObject o = new JsonObject();
        o.addProperty("kind", "fatal");
        o.addProperty("reason", reason);
        writeLine(o);
    }

    void event(JsonObject event) {
        JsonObject o = new JsonObject();
        o.addProperty("kind", "event");
        o.add("event", event);
        writeLine(o);
    }

    // --- canonical vt_model Event constructors (tag = "type") ---

    void orderAccepted(String coid, String venueOrderId, long ts) {
        JsonObject e = evt("OrderAccepted", coid, ts);
        e.addProperty("venue_order_id", venueOrderId);
        event(e);
    }

    void orderRejected(String coid, String reason, long ts) {
        JsonObject e = evt("OrderRejected", coid, ts);
        e.addProperty("reason", reason);
        event(e);
    }

    void orderCanceled(String coid, String reason, long ts) {
        JsonObject e = evt("OrderCanceled", coid, ts);
        e.addProperty("reason", reason);
        event(e);
    }

    /**
     * Dual-publish contract (mirrors the crypto venue mappers): a fill emits BOTH the bare
     * FillEvent (the core Account folds it into position/PnL — it is the SOLE writer of
     * position/realized PnL, reached only via this lane) AND the wrapping OrderFilled/
     * OrderPartiallyFilled (the order FSM applies it), in that order, both carrying the same
     * fill. full=true -> OrderFilled wrap, else OrderPartiallyFilled. The two share the fill's
     * trade_id so the core's separate Account/FSM dedup sets each key correctly on a replay.
     */
    void orderFilled(boolean full, String coid, JsonObject fill, long ts) {
        // Bare FillEvent lane: the event object IS the fill fields, tagged "FillEvent"
        // (the Event enum's wire tag for the bare-fill variant). deepCopy so tagging it
        // does not mutate the fill embedded in the wrap below.
        JsonObject bare = fill.deepCopy();
        bare.addProperty("type", "FillEvent");
        event(bare);
        // Wrapping lane: the order FSM.
        JsonObject e = evt(full ? "OrderFilled" : "OrderPartiallyFilled", coid, ts);
        e.add("fill", fill);
        event(e);
    }

    /**
     * The venue-authoritative net position line (kind "position", netting-truth law A7): signed
     * size in UNITS + the signed-amount-weighted average entry of the remaining orders
     * ({@link NetPosition}). Emitted after every fill so the Rust side can re-anchor its blended
     * fold to the per-order attribution this venue actually realized. Additive: an old Rust
     * reader treats the unknown kind as an unparseable line and logs-and-skips.
     */
    void positionState(String symbol, double sizeUnits, double avgPx, long ts) {
        JsonObject o = new JsonObject();
        o.addProperty("kind", "position");
        o.addProperty("symbol", symbol);
        o.addProperty("size", sizeUnits);
        o.addProperty("avg_px", avgPx);
        o.addProperty("ts", ts);
        writeLine(o);
    }

    private static JsonObject evt(String type, String coid, long ts) {
        JsonObject e = new JsonObject();
        e.addProperty("type", type);
        e.addProperty("client_order_id", coid);
        e.addProperty("ts", ts);
        return e;
    }
}
